.github/workflows/test.yaml

```diff
@@ -0,0 +1,131 @@
+
+      - name: Log into container registry
+        run: echo ${{ secrets.DockerHubToken }} | docker login --username ${DOCKER_USER} --password-stdin
+
+      - name: build and push runner image
+        env:
+          # ${{ github.head_ref }} is empty for non-pull requests, which is
+          # exactly what we want on master.
+          RUNNER_IMAGE_TAG_PREFIX: ${{ github.head_ref }}
+        run: RUNNER_IMAGE=${DOCKER_ORG}/k8s-e2e-test-runner make runner-push
+
+      - name: Build and push plugin image
+        env:
+          VERSION: ${{ github.head_ref }}
+        run: |
+          VERSION="${VERSION:-latest}"
+          DOCKER_REPO=${DOCKER_ORG}/do-csi-plugin-dev make compile build push
+
+  e2e-test:
+    runs-on: ubuntu-18.04
+    needs: push-images
+    strategy:
+      matrix:
+        kube-release: ['1.16', '1.15', '1.14']
+
+    steps:
+      - name: checkout
+        uses: actions/checkout@master
+
+      - name: Go setup
+        uses: actions/setup-go@v1
+        with:
+          go-version: '1.13.5'
+
+      - name: Install kustomize
+        env:
+          KUSTOMIZE_VERSION: '3.5.3'
+        run: curl -fL https://github.com/kubernetes-sigs/kustomize/releases/download/kustomize%2Fv${KUSTOMIZE_VERSION}/kustomize_v${KUSTOMIZE_VERSION}_linux_amd64.tar.gz | sudo tar xzv -C /usr/local/bin kustomize
+
+      - name: Run end-to-end tests
+        env:
+          DIGITALOCEAN_ACCESS_TOKEN: ${{ secrets.CSIDigitalOceanAccessToken }}
+          # ${{ github.head_ref }} will be empty for pull requests and non-empty
+          # for pushes. Handle the cases in "run" below to parameterize our e2e
+          # test invocation accordingly.
+          BRANCH: ${{ github.head_ref }}
+          # Use less ginkgo nodes to avoid running into the DO API rate limit.
+          # The upstream end-to-end tests are quite demanding in terms of API
+          # request volume, and we run them concurrently.
+          NUM_GINKGO_NODES: "8"
+        run: |
+          NAME_SUFFIX="${BRANCH}"
+          TAG="${BRANCH}"
+          if [[ $BRANCH != "" ]]; then
+            RUNNER_IMAGE_TAG_PREFIX=${BRANCH}-
+          else
+            NAME_SUFFIX=master
+            TAG=latest
+          fi
+          TIMEOUT=60m make test-e2e E2E_ARGS="-ginkgo-nodes ${NUM_GINKGO_NODES} -driver-image ${DOCKER_ORG}/do-csi-plugin-dev:${TAG} -runner-image ${DOCKER_ORG}/k8s-e2e-test-runner:${RUNNER_IMAGE_TAG_PREFIX}latest -name-suffix ${NAME_SUFFIX} ${{ matrix.kube-release }}"
+
+  tag-new-master-image:
+    runs-on: ubuntu-18.04
+    if: github.ref == 'refs/heads/master'
+    needs: e2e-test
+
+    steps:
+      - name: checkout
+        uses: actions/checkout@master
+
+      - name: Log into container registry
+        run: echo ${{ secrets.DockerHubToken }} | docker login --username ${DOCKER_USER} --password-stdin
+
+      - name: tag new CSI plugin master image
+        run: |
+          docker pull ${DOCKER_ORG}/do-csi-plugin-dev:latest
```

Comment:
Do we need latest if we just build as `master`? I guess I could see an argument in terms of semantics but they will always be in-sync as far I can tell?

Reply:
My take on this is that `master` is the final product of our CI run that is known to have passed all tests successfully; other than for this final step in our CI, it is not meant to be otherwise updated but consumed only.

`latest` on the other hand is WIP until we have successfully completed the final step in our CI.  It's an implementation detail that could contain broken code until CI has finally assured correctness.

To put it in more practical terms: if we used a single `master` tag only, we could end up in a situation where we are updating our single image tag in the _Build and push plugin image_ step (as the PR does today) only to find it broken in the `Run end-to-end tests` step; at that point, our single `master` image broke, and so will potential (external) consumers.

Does my train of thought make sense?

Reply:
Ahh, I see what you are saying, that makes sense to me. I guess the only thing that comes to mind as a minor concern is if we have two PRs merged near one another, then the builds will be competing for contents of that tag.

I'm not sure if it makes sense to use a commit hash based approach for this intermediate build and then cleanup later on versus clashing on master. I will leave that up to you, I think for the most part we shouldn't run into issues.

Reply:
Hmm yeah that's a fair point. What I'm not sure is if two consecutive master builds would be able to compete. Does the latter perhaps cancel the former build? I haven't really tested this to be honest.

I'm inclined to giving this a try as it stands now and adjust if it turns out to be problematic.